extensions/validation.js

```diff
@@ -0,0 +1,46 @@
+import Ajv from 'ajv';
+import axios from 'axios';
+import { dirname, join } from 'path';
+import { fileURLToPath } from 'url';
+import { promises as fs, readdirSync, statSync } from 'fs';
+import iriFormats from 'stac-node-validator/iri.js';
+const __dirname = dirname(fileURLToPath(import.meta.url));
+
+/**
+ * function passed in to Ajv instance which allows us to load schemas from a url at run time.
+ */
+async function loadSchema(uri) {
+  let json;
+  try {
+    let response = await axios.get(uri);
+    json = response.data;
+  } catch (error) {
+    throw new Error(`-- Schema at '${uri}' not found. Please ensure all entries in 'stac_extensions' are valid.`);
+  }
+  return json;
+}
+
+/**
+ * function which will preload the Ajv instance with schemas from each directory beneath the 'extensions' directory.
+ */
+async function populateSchemas(path) {
+  let schemaList = readdirSync(path).filter(function (file) {
```

Comment:
This is a weird mix of async and sync code.

I would write this something like

```
const fileList = await fs.promises.readdir('extensions', { withFileTypes: true});

for (const file of fileList) {
    if (!file.isDirectory()) continue;
    const schema = path.join('extensions', file.path, 'schema.json');
    await loadSchema(schema);
}
```

We use a file system abstraction `@chunkd/fs` in quite a few places it makes it easy to list and read files from s3/http and local file systems

so you can `fs.read("https://foo/bar")` or `fs.read("/home/blacha/foo/bar")` which means you can remove the extra axios logic for handling http links

it also simplifies the listing/reading process

```
// recursively list all files inside of a folder
for await (const  fileName of fs.list(filePath)) {
   if (!fileName.endsWitch("schema.json")) continue
   const data = await fs.read(fileName);
}
```